File: app/source/js/models/QuestionResponse.js
/**
 * QuestionResponse Model matching the plain rest.QuestionResponse type.
 */

import Model from './Model';

/**
 * Describes a response given by a user to a question, that may include user
 * input, a selected option or both.
 * @typedef {Object} rest.QuestionResponse
 * @property {number} [optionID] Unique identifier for the option in the question
 * if one was choosen. On questions with a single option for user input, this
 * property is not required.
 * @property {string} [option] Text used to label the optionID selected (is a
 * copy of the text displayed to the user when answering the question, only
 * needed when there is an optionID).
 * @property {string} [userInput] The input provided by the user, usually in a
 * free text box, but can be other type like number, datetime
 * choosing this option, or null when no input (just choosing the option is enough).
 * The values allowed matches the values allowed by the attribute 'type' of
 * the html 'input' element.
 */

export default class QuestionResponse {
    constructor(values) {
        Model(this);

        this.model.defProperties({
            optionID: null,
            option: null,
            userInput: null
        }, values);
    }
}

File: web/App_Code/LcRest/QuestionResponse.cs
namespace LcRest
{
    /// <summary>
    /// Represents a single response to a question
    /// </summary>
    public class QuestionResponse
    {
        public int? optionID;
        public string option;
        public string userInput;

        public QuestionResponse() { }

        public static QuestionResponse FromDB(dynamic record)
        {
            if (record == null) return null;
            return new QuestionResponse
            {
                optionID = record.optionID,
                option = record.option,
                userInput = record.userInput
            };
        }
    }
}
